cmd/kubeapps-apis/plugins/resources/v1alpha1/server.go

```diff
@@ -130,34 +135,164 @@ func (s *Server) GetResources(r *v1alpha1.GetResourcesRequest, stream v1alpha1.R
 		if err != nil {
 			return status.Errorf(codes.Internal, "unable to parse group version from %q: %s", ref.ApiVersion, err.Error())
 		}
 		gvk := groupVersion.WithKind(ref.Kind)
 		// TODO(minelson): Find alternative to UnsafeGuessKindToResource.
+		// Looks to involve restmapper.NewDeferredDiscoveryRESTMapper(discovery.NewDiscoveryClient(c))
+		// See https://pkg.go.dev/k8s.io/client-go/restmapper#NewDeferredDiscoveryRESTMapper
 		gvr, _ := meta.UnsafeGuessKindToResource(gvk)
-		resource, err := dynamicClient.Resource(gvr).Namespace(namespace).Get(stream.Context(), ref.GetName(), metav1.GetOptions{})
-		if err != nil {
-			return status.Errorf(codes.Internal, "unable to get resource referenced by %+v: %s", ref, err.Error())
+
+		if !r.GetWatch() {
+			resource, err := dynamicClient.Resource(gvr).Namespace(namespace).Get(stream.Context(), ref.GetName(), metav1.GetOptions{})
+			if err != nil {
+				return status.Errorf(codes.Internal, "unable to get resource referenced by %+v: %s", ref, err.Error())
+			}
+
+			err = sendResourceData(ref, resource, stream)
+			if err != nil {
+				return err
+			}
+
+			continue
 		}
 
-		resourceBytes, err := json.Marshal(resource)
+		watcher, err := dynamicClient.Resource(gvr).Namespace(namespace).Watch(stream.Context(), metav1.ListOptions{})
 		if err != nil {
-			return status.Errorf(codes.Internal, "unable to marshal json for resource: %s", err.Error())
+			return status.Errorf(codes.Internal, "unable to watch resource %v", ref)
 		}
-		stream.Send(&v1alpha1.GetResourcesResponse{
+		watchers = append(watchers, &ResourceWatcher{
 			ResourceRef: ref,
-			Manifest: &anypb.Any{
-				Value: resourceBytes,
-			},
+			Watcher:     watcher,
 		})
 	}
 
+	// If we're not watching, we're done.
+	if watchers == nil {
+		return nil
+	}
+
+	// Otherwise merge the watchers into a single resourceWatcher and stream the
+	// data as it arrives.
+	resourceWatcher := mergeWatchers(watchers)
+	for e := range resourceWatcher.ResultChan() {
+		sendResourceData(e.ResourceRef, e.Object, stream)
+	}
+
 	return nil
 }
 
+// sendResourceData just DRYs up this functionality shared between requests to
+// watch or get resources.
+func sendResourceData(ref *pkgsGRPCv1alpha1.ResourceRef, obj interface{}, s v1alpha1.ResourcesService_GetResourcesServer) error {
+	resourceBytes, err := json.Marshal(obj)
+	if err != nil {
+		return status.Errorf(codes.Internal, "unable to marshal json for resource: %s", err.Error())
+	}
+	s.Send(&v1alpha1.GetResourcesResponse{
+		ResourceRef: ref,
+		Manifest: &anypb.Any{
+			Value: resourceBytes,
+		},
+	})
+
+	return nil
+}
+
+// ResourceEvent embeds a watch.Event and adds the resource ref.
+type ResourceEvent struct {
+	watch.Event
+	ResourceRef *pkgsGRPCv1alpha1.ResourceRef
+}
+
+// mergeWatchers returns a single watcher for many.
+// Inspired by the fan-in merge at https://go.dev/blog/pipelines
+func mergeWatchers(watchers []*ResourceWatcher) *MultiResourceWatcher {
```

Comment:
This is really interesting! Thanks for the link; I'll take a closer look at it (channels and mastering goroutines are still in my todo list of GO things to learn :P)